File: bcs-ui/backend/iam/open_apis/providers/project.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from iam.collection import FancyDict
from iam.resource.provider import ListResult, ResourceProvider
from iam.resource.utils import Page

from backend.container_service.projects.base import list_projects, page_query_projects

from .utils import get_system_token


class ProjectProvider(ResourceProvider):
    """项目 Provider"""

    def list_instance(self, filter_obj: FancyDict, page_obj: Page, **options) -> ListResult:
        data = page_query_projects(get_system_token(), limit=page_obj.limit, offset=page_obj.offset)
        results = [{'id': p['project_id'], 'display_name': p['project_name']} for p in data['results']]
        return ListResult(results=results, count=data['count'])

    def fetch_instance_info(self, filter_obj: FancyDict, **options) -> ListResult:
        query_params = {'project_ids': ','.join(filter_obj.ids)}
        projects = list_projects(get_system_token(), query_params)
        results = [
            {
                'id': p['project_id'],
                'display_name': p['project_name'],
                '_bk_iam_approver_': [p['creator'], p.get('updater')],
            }
            for p in projects
        ]
        return ListResult(results=results, count=len(results))

    def list_instance_by_policy(self, filter_obj: FancyDict, page_obj: Page, **options) -> ListResult:
        # TODO 确认基于实例的查询是不是就是id的过滤查询
        return ListResult(results=[], count=0)

    def list_attr(self, **options) -> ListResult:
        return ListResult(results=[], count=0)

    def list_attr_value(self, filter_obj: FancyDict, page_obj: Page, **options) -> ListResult:
        return ListResult(results=[], count=0)

    def search_instance(self, filter_obj: FancyDict, page_obj: Page, **options) -> ListResult:
        """支持模糊搜索项目名"""
        data = page_query_projects(
            get_system_token(),
            limit=page_obj.limit,
            offset=page_obj.offset,
            query_params={'search_name': filter_obj.keyword},
        )
        if data and data['results']:
            results = [{'id': p['project_id'], 'display_name': p['project_name']} for p in data['results']]
            return ListResult(results=results, count=data['count'])
        return ListResult(results=[], count=0)

File: bcs-services/bcs-user-manager/app/user-manager/v1http/iam/project.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package iam

import (
	"context"
	"sync"

	"github.com/TencentBlueKing/iam-go-sdk/resource"

	"github.com/Tencent/bk-bcs/bcs-services/bcs-user-manager/app/pkg/component"
	blog "github.com/Tencent/bk-bcs/bcs-services/bcs-user-manager/app/pkg/log"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-user-manager/app/pkg/utils"
	util "github.com/Tencent/bk-bcs/bcs-services/bcs-user-manager/app/utils"
)

// ProjectProvider is a project provider
type ProjectProvider struct {
}

func init() {
	dispatcher.RegisterProvider(Project, ProjectProvider{})
}

// ListAttr implements the list_attr
func (p ProjectProvider) ListAttr(req resource.Request) resource.Response {
	return resource.Response{
		Code: 0,
		Data: []interface{}{},
	}
}

// ListAttrValue implements the list_attr_value
func (p ProjectProvider) ListAttrValue(req resource.Request) resource.Response {
	return resource.Response{
		Code: 0,
		Data: ListResult{Count: 0, Results: []interface{}{}},
	}
}

// ListInstance implements the list_instance
func (p ProjectProvider) ListInstance(req resource.Request) resource.Response {
	offset := req.Page.Offset / req.Page.Limit
	tenantID := req.Header.Get(util.HeaderTenantID)
	result, err := component.QueryProjects(req.Context, tenantID, req.Page.Limit, offset, nil)
	if err != nil {
		return resource.Response{
			Code:    SystemErrCode,
			Message: err.Error(),
		}
	}
	results := make([]interface{}, 0)
	for _, r := range result.Results {
		ins := Instance{r.ProjectID, combineNameID(r.Name, r.GetProjectCode()), nil}
		results = append(results, ins)
	}
	return resource.Response{
		Code: 0,
		Data: ListResult{Count: result.Total, Results: results},
	}
}

// FetchInstanceInfo implements the fetch_instance_info
func (p ProjectProvider) FetchInstanceInfo(req resource.Request) resource.Response {
	filter := convertFilter(req.Filter)
	if len(filter.IDs) == 0 {
		return resource.Response{
			Code:    NotFoundCode,
			Message: "ids is empty",
			Data:    []interface{}{},
		}
	}

	nsChan := make(chan Instance, len(filter.IDs))
	results := make([]interface{}, 0)
	wg := sync.WaitGroup{}
	for _, v := range filter.IDs {
		wg.Add(1)
		go func(id string) {
			defer wg.Done()
			ctx := context.Background()
			ctx = context.WithValue(ctx, utils.ContextValueKeyRequestID, utils.GetRequestIDFromContext(req.Context))
			p, err := component.GetProject(ctx, id)
			if err != nil {
				blog.Log(ctx).Errorf("get project %s failed, err %s", id, err.Error())
				return
			}
			nsChan <- Instance{p.ProjectID, combineNameID(p.Name, p.GetProjectCode()), SplitString(p.Managers)}
		}(v)
	}
	wg.Wait()
	close(nsChan)

	for r := range nsChan {
		results = append(results, r)
	}
	return resource.Response{
		Code: 0,
		Data: results,
	}
}

// ListInstanceByPolicy implements the list_instance_by_policy
func (p ProjectProvider) ListInstanceByPolicy(req resource.Request) resource.Response {
	return resource.Response{
		Code: 0,
		Data: ListResult{Count: 0, Results: []interface{}{}},
	}
}

// SearchInstance implements the search_instance
func (p ProjectProvider) SearchInstance(req resource.Request) resource.Response {
	filter := convertFilter(req.Filter)
	tenantID := req.Header.Get(util.HeaderTenantID)
	offset := req.Page.Offset / req.Page.Limit
	params := map[string]string{"searchName": filter.Keyword}
	result, err := component.QueryProjects(req.Context, tenantID, req.Page.Limit, offset, params)
	if err != nil {
		return resource.Response{
			Code:    SystemErrCode,
			Message: err.Error(),
		}
	}
	results := make([]interface{}, 0)
	for _, r := range result.Results {
		ins := Instance{r.ProjectID, combineNameID(r.Name, r.GetProjectCode()), nil}
		results = append(results, ins)
	}
	return resource.Response{
		Code: 0,
		Data: ListResult{Count: result.Total, Results: results},
	}
}

// FetchInstanceList implements the fetch_instance_list
func (p ProjectProvider) FetchInstanceList(req resource.Request) resource.Response {
	return resource.Response{
		Code:    -1,
		Message: "not implemented",
	}
}

// FetchResourceTypeSchema implements the fetch_resource_type_schema
func (p ProjectProvider) FetchResourceTypeSchema(req resource.Request) resource.Response {
	return resource.Response{
		Code:    -1,
		Message: "not implemented",
	}
}
